### PR TITLE
RFC-0151_TransactionProtocol

### PR DESCRIPTION
**Pull request by CjS77**
## Description

This document describes the transaction protocol for peer-to-peer Tari payments using the 
Mimblewimble protocol. It also
considers some attacks that may be launched against the protocol and offers some discussion around those attacks and
potential alternatives to the protocol.

The goal is to describe a transaction protocol that
* permits multiple recipients,
* preserves privacy regarding how many parties are involved in the transaction,
* is secure against all reasonable attacks.

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [x] Documentation

## Checklist:

* [x] I'm merging against the `development` branch
* [ ] I ran `cargo-fmt --all` before pushing
* [ ] My change requires a change to the documentation.
* [x] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

### CODE REVIEW

File: RFC/src/RFC-0151_TransactionProtocol.md [hunk 1]
```diff
@@ -0,0 +1,255 @@
+# TransactionProtocol
+
+## Transaction protocol
+
+![status: draft](theme/images/status-draft.svg)
+
+**Maintainer(s)**: <Maintainer name> <github username>
```
Comment (neonknight64) [inline RFC/src/RFC-0151_TransactionProtocol.md]: Missing maintainer

File: RFC/src/RFC-0151_TransactionProtocol.md [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+        note left of Sender: CollectingCommitments
+        note right of Receivers: SendingCommitment
+        Receivers-->>-Sender: [tx_id, H(Ri||Pi)]
+     deactivate Sender
+# Receiving Outputs
+     activate Sender
+        Sender-->>+Receivers: [tx_id, [Hashed commitments]]
+        note left of Sender: CollectingPubKeys
+        note right of Receivers: SendingOutput
+        Receivers-->>-Sender: [tx_id, Pi, Ri, C_i, RP_i]
```
Comment (philipr-za) [inline RFC/src/RFC-0151_TransactionProtocol.md]: I am not certain what C_i and RP_i are in this step. Guessing they are analogues of terms from MuSig but it is not clear from this context.

File: RFC/src/RFC-0151_TransactionProtocol.md [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+    deactivate Sender
+    alt inflation_error()?
+        Sender--XReceivers: [tx-id, failed]
+    end
+# Signature collection
+    activate Sender
+        Sender-->>+Receivers: [tx_id, [Rs, Ri] [Xs, Pi]]
```
Comment (philipr-za) [inline RFC/src/RFC-0151_TransactionProtocol.md]: What is the value of `i` in the the Ri and Pi here. Are you sending back each individuals Receivers nonce and Public key or are you sending a collection of all of them?

File: RFC/src/RFC-0151_TransactionProtocol.md [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+        note left of Sender: CollectingSignatures
+        note right of Receivers: Signing
+        Receivers-->>Receivers: validate nonce and pubkey sum
```
Comment (philipr-za) [inline RFC/src/RFC-0151_TransactionProtocol.md]: Do you mean the Sender's nonce, Rs, and excess, Xs, (is this what you mean by public key sum) are validated against the hash sent in the first step, H(Rs||Xs)?

Comment (philipr-za) [reply RFC/src/RFC-0151_TransactionProtocol.md]: Or do you mean checking all the respective nonces and keys against the the individual H(Ri||Pi) terms? It's the `pubkey sum` that is confusing me.

File: RFC/src/RFC-0151_TransactionProtocol.md [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+calculate these sums themselves.
+
+Unfortunately, as we discuss below, it seems that it's not possible using any known scheme that satisfies both this
+privacy goal while achieving the desired privacy level.
+
+
+### The issue with multiple senders
+
+To increase privacy, the public excess values are [offset] by a constant random value. The choice of this value, as well
+as fee selection can only be set once per transaction. The privilege of selecting these values is generally bestowed on
```
Comment (neonknight64) [inline RFC/src/RFC-0151_TransactionProtocol.md]:
```suggestion
as fee selection, can only be set once per transaction. The privilege of selecting these values is generally bestowed on
```

File: RFC/src/RFC-0151_TransactionProtocol.md [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+    participant Sender
+    participant Receivers
+# Sender Initializes the transaction
+    activate Sender
+        Sender-->>Sender: initialize
+    deactivate Sender
+# Sender transmits initial data to all receivers
+    activate Sender
+        Sender-->>+Receivers: [tx_id, amt_i,H(Rs||Xs),m]
+        note left of Sender: CollectingCommitments
```
Comment (neonknight64) [inline RFC/src/RFC-0151_TransactionProtocol.md]: In the rendered figure, "CollectingCommitments" doesn't fit into its yellow box.

File: RFC/src/RFC-0151_TransactionProtocol.md [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+* preserves privacy regarding how many parties are involved in the transaction,
+* is secure against all reasonable attacks.
+
+## Related RFCs
+
+* [RFC-0100: The Base Layer](RFC-0100_BaseLayer.md)
+
+## Description
+
+The Tari base layer is built using [Mimblewimble], which requires all parties involved in a Tari transfer must interact
```
Comment (hansieodendaal) [inline RFC/src/RFC-0151_TransactionProtocol.md]:
```suggestion
The Tari base layer is built using [Mimblewimble], which requires that all parties involved in a Tari transfer must interact
```

File: RFC/src/RFC-0151_TransactionProtocol.md [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+and _m_ is the additional metadata for the transaction. \\(\Sigma P_i\\) is the value of the (pre-offset) excess that is
+stored in the transaction kernel.
+
+Notice that every signing party needs to know the sum of all the nonces and public spending keys. This suggests that
+every party knows how many parties are involved in the transaction, which is not an ideal privacy scenario. It would be
+preferable if a secure scheme could be found where each recipient interacts only with the sender and does not need to
+calculate these sums themselves.
+
+Unfortunately, as we discuss below, it seems that it's not possible using any known scheme that satisfies both this
+privacy goal while achieving the desired privacy level.
```
Comment (hansieodendaal) [inline RFC/src/RFC-0151_TransactionProtocol.md]:
```suggestion
privacy goal while achieving the desired security level.
```

File: RFC/src/RFC-0151_TransactionProtocol.md (hunk 3 above)
Comment (CjS77) [reply RFC/src/RFC-0151_TransactionProtocol.md]: We're sending all of them

File: RFC/src/RFC-0151_TransactionProtocol.md (hunk 6 above)
Comment (CjS77) [reply RFC/src/RFC-0151_TransactionProtocol.md]: I know. Blame mermaid, not me :)

**Pull request merged**